### PR TITLE
sql: Adding support for bytea hex format

### LINKED ISSUE #15880: sql/parser: scanner does not support e'\\xDEADBEEF` binary input format
The test case in https://gist.github.com/davestgermain/7eae5c1b148a1e0dfa09d3818fdd1cfb (run with python 3.6 and psycopg2 2.7.1) shows that binary data is getting returned as a SQL hex-encoded string (`\\xffd8ffe00010`), and the python psycopg2 driver doesn't know how to decode that back to the raw bytes. I'm not sure at what level things are going wrong, but the test works as expected when run against postgresql.

### PR DESCRIPTION
**Pull request by m-schneider**
Before this change escaped bytea hex data would be saved as an escaped
string.

After this change binary data with the [e,E]'\\[X,x] prefix will be
parsed as binary.

Closes #15880

### CODE REVIEW

**Comment by jordanlewis** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/scan.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrgQK2jMzHtApK-Z5ew:-KrgQK2kLj8tipnVoEfN:b-cxcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e46105019e9a509a50065aaacc30c088e6a563/pkg/sql/parser/scan.go#L259)):*
> ```Go
> 			s.pos++
> 			// We're checking for the escaped version of a BYTEA represented in hex.
> 			if s.peek() == '\\' && unicode.ToLower(rune(s.peekN(2))) == 'x' {
> ```

maybe just explicitly check for 'x' and 'X' like we do a few lines below in the case statement to skip the unicode toLower?

---

*[pkg/sql/parser/scan.go, line 706 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrgR3vTVxk_PbfH9l24:-KrgR3vTVxk_PbfH9l25:br9oddo) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e46105019e9a509a50065aaacc30c088e6a563/pkg/sql/parser/scan.go#L706)):*
> ```Go
> 						tmp = s.in[s.pos-1:]
> 					}
> 					if unicode.ToLower(rune(s.peekN(1))) == 'x' && decodeHex {
> ```

same here, skip the toLower if possible?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

**Comment by m-schneider** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/scan.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrgQK2jMzHtApK-Z5ew:-KrgSndDjSX1e5R0sK_b:b-80tlqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e46105019e9a509a50065aaacc30c088e6a563/pkg/sql/parser/scan.go#L259)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

maybe just explicitly check for 'x' and 'X' like we do a few lines below in the case statement to skip the unicode toLower?
</blockquote></details>

Sure!

---

*[pkg/sql/parser/scan.go, line 706 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrgR3vTVxk_PbfH9l24:-KrgSqapbLQXCuNSP6fG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e46105019e9a509a50065aaacc30c088e6a563/pkg/sql/parser/scan.go#L706)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

same here, skip the toLower if possible?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

**Comment by jordanlewis** [comment]
:lgtm: once tests pass - this is not as bad as you made it sound!

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-:-KrgVNy2ijMJDCykngto:bdbuol2)*

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/parser/scan.go, line 707 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrhlfaqcK-AAYFiIne5:-KrhlfaqcK-AAYFiIne6:b5a5f8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fcb3ae3720a911d04758b1ea8bbd971c368309/pkg/sql/parser/scan.go#L707)):*
> ```Go
> 					}
> 					lookAhead := s.peekN(1)
> 					if lookAhead == 'x' && decodeHex || lookAhead == 'X' {
> ```

This might be a stupid question, but what is the `&& decodeHex` doing here? It doesn't seem to be necessary for the test cases you've added. And I'm having a hard time understanding why `decodeHex` only applies to `lookAhead == 'x'`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

**Comment by knz** [comment]
Huh wait Houston (I think) we have a problem.

The pg docs are a absolutely contradictory about this:

- https://www.postgresql.org/docs/9.6/static/sql-syntax-lexical.html#SQL-SYNTAX-STRINGS-ESCAPE
  Says that `E'\x4142'` is equivalent to "A42"

- https://www.postgresql.org/docs/9.6/static/datatype-binary.html#AEN5757
  Says that `E'\x4142'` is equivalent to "AB"

There are two possible situations:

- either one of the two docs is plainly incorrect. Then you must look at the source code of postgres, do the same in your code, and link to the pg code (and outline the inconsistency in the docs) both in the commit message and around the code in comments.

- both docs are correct, which implies that postgres only parses the escapes *after* type checking (i.e. it changes the behavior depending on whether typing decides for BYTEA or STRING, and perhaps it "prefers" typing a string leteral as BYTEA if the string happens to contain the substring "\x"). In that case you also need to do that.

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-:-Krjrl10pF8J79NzNXqW:byn4c8q)*

**Comment by knz** [comment]
@bdarnell had you noticed that pg is inconsistent on this? what do you think is least surprising?

**Comment by knz** [comment]
@nvanbenschoten I would welcome your opinion on this too

**Comment by knz** [comment]
Also, another aspect: I have looked at #15880 and I do not understand exactly the relationship between this PR and that bug report. In #15880 I believe there are two separate issues, and only one is addressed in this PR:

- the scanner may or may not know how to parse string literals properly (see discussion above)
- our pgwire may or may not know how to output byte arrays properly (would need to check the pg specs and/or what a pg server says on pgwire under the same circcumstances as outlined in #15880)

**Comment by jordanlewis** [comment]
@knz, there is no inconsistency - the number of backslashes matters, and as @m-schneider and I realized yesterday, this current patch is flawed - the transformation here should only be applied on cast to bytes.

Observe pg:

```
jordan=# select e'\\xdeadbeez'::bytea;
ERROR:  invalid hexadecimal digit: "z"
LINE 1: select e'\\xdeadbeez'::bytea;
               ^
jordan=# select e'\\xdeadbeez';
  ?column?
------------
 \xdeadbeez
(1 row)
```

This illustrates that pg only does the transformation at cast time. Additionally, it implies that the input string `e'\\xdeadbeef'` without a cast returns the string literal `"\xdeadbeez"`.

The other syntax you mention uses *single backslashes* to escape 2-character byte sequences. As in your example:

```
jordan=# select e'\x4142';
 ?column?
----------
 A42
(1 row)
```

We already support that syntax correctly:

```
root@:26257/> select e'\x4142';
+-------+
| 'A42' |
+-------+
| A42   |
+-------+
(1 row)
```

In conclusion, everything is fine - the approach of this PR will be changing the string->bytes cast mechanism to detect an `\x` (input via escape string by `e'\\x`) and treat the rest of the string as hex digits.

The linked issue will be solved by this. We already know how to output bytes to pgwire correctly.

**Comment by knz** [comment]
All right, then I'll look again when the code is updated. Let me know.

**Comment by m-schneider** [comment]
I've updated the code to take into account the behavior that we discussed with Jordan.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/parser/scan.go, line 707 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-KrhlfaqcK-AAYFiIne5:-KrloSGQ7uph2qnyfFmj:buu9hd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fcb3ae3720a911d04758b1ea8bbd971c368309/pkg/sql/parser/scan.go#L707)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This might be a stupid question, but what is the `&& decodeHex` doing here? It doesn't seem to be necessary for the test cases you've added. And I'm having a hard time understanding why `decodeHex` only applies to `lookAhead == 'x'`.
</blockquote></details>

After looking into this more with Jordan I took the changes out of the scanner. Before I called the scanStringOrHex function for escaped strings with decodeHex if their prefixes were e'\\x, which I agree is a bit too hacky,. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-:-KrloBEZKYkNjCBADCUQ:bi5uttn)*

**Comment by petermattis** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 205 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-Krms6Q7pamw9SGfdhoU:-Krms6Q7pamw9SGfdhoV:bvymvhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/07565967f6435213f53f8c0a1f6d4d4cc9fb8bc4/pkg/sql/parser/datum.go#L205)):*
> ```Go
> // ParseDByte parses a string representation of hex encoded binary data.
> func ParseDByte(s string) (*DBytes, error) {
> 	re := regexp.MustCompile(`^\\[x,X]`)
> ```

Premature optimization of the root of all evil, yet we probably don't want to parse a regex on every call to `ParseDByte`. And regexes are usually not the best idea for performance sensitive code. This could be written as:

```
  if len(s) >= 2 && (s[0] == '\\' && (s[1] == 'x' || s[1] == 'X')) {
    ...
  }
```

Definitely a bit more complex, but there is a reason for using this style in the scanner and I think the same rationale applies here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

**Comment by petermattis** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 205 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-Krms6Q7pamw9SGfdhoU:-KrmtWqzy7wSG1BYSDuE:bfhq53o) ([raw file](https://github.com/cockroachdb/cockroach/blob/07565967f6435213f53f8c0a1f6d4d4cc9fb8bc4/pkg/sql/parser/datum.go#L205)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Premature optimization of the root of all evil, yet we probably don't want to parse a regex on every call to `ParseDByte`. And regexes are usually not the best idea for performance sensitive code. This could be written as:

```
  if len(s) >= 2 && (s[0] == '\\' && (s[1] == 'x' || s[1] == 'X')) {
    ...
  }
```

Definitely a bit more complex, but there is a reason for using this style in the scanner and I think the same rationale applies here. 
</blockquote></details>

Also, I just realized that regex matches the string `"\\,"`. Probably not what you intended. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

**Comment by knz** [comment]
LGTM modulo the change asked for by Peter. I concur with him.

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-:-KroGKIrhimAFMqdwrN3:b-asg7nh)*

**Comment by m-schneider** [comment]
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/datum.go, line 205 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693#-Krms6Q7pamw9SGfdhoU:-Ks47yXmj1Kd_VAej_Ek:b-uo2vq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/07565967f6435213f53f8c0a1f6d4d4cc9fb8bc4/pkg/sql/parser/datum.go#L205)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Also, I just realized that regex matches the string `"\\,"`. Probably not what you intended. 
</blockquote></details>

Ok, I'll keep that in mind in for the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17693)*

File: pkg/sql/parser/datum.go [hunk 1]
```diff
@@ -199,6 +200,18 @@ func ParseDBool(s string) (*DBool, error) {
 	return MakeDBool(DBool(b)), nil
 }
 
+// ParseDByte parses a string representation of hex encoded binary data.
+func ParseDByte(s string) (*DBytes, error) {
+	if len(s) > 2 && (s[0] == '\\' && (s[1] == 'x' || s[1] == 'X')) {
+		hexstr, err := hex.DecodeString(s[2:])
+		if err != nil {
+			return NewDBytes(DBytes(s)), nil
```
Comment (jordanlewis) [inline pkg/sql/parser/datum.go]: Sorry I was slow to look at this -- this does introduce a minor incompatibility with Postgres, which will return an error if you send it invalid hex - e.g. `e'\\xthisisnthex'`. I'm not sure how much this matters, but I feel like we might as well fix it anyway.
